fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FailSoftMapCodec.java

```diff
@@ -0,0 +1,129 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.dimension;
+
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableMap;
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.Lifecycle;
+import com.mojang.serialization.MapLike;
+import com.mojang.serialization.codecs.BaseMapCodec;
+import com.mojang.serialization.codecs.UnboundedMapCodec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Has the same functionality as {@link UnboundedMapCodec}.
+ * But it will fail-soft when an entry cannot be deserialized.
+ */
+public class FailSoftMapCodec<K, V> implements BaseMapCodec<K, V>, Codec<Map<K, V>> {
+	public static final Logger LOGGER = LoggerFactory.getLogger("FailSoftMapCodec");
+
+	private final Codec<K> keyCodec;
+	private final Codec<V> elementCodec;
+
+	public FailSoftMapCodec(final Codec<K> keyCodec, final Codec<V> elementCodec) {
+		this.keyCodec = keyCodec;
+		this.elementCodec = elementCodec;
+	}
+
+	@Override
+	public Codec<K> keyCodec() {
+		return keyCodec;
+	}
+
+	@Override
+	public Codec<V> elementCodec() {
+		return elementCodec;
+	}
+
+	@Override
+	public <T> DataResult<Pair<Map<K, V>, T>> decode(final DynamicOps<T> ops, final T input) {
+		return ops.getMap(input).setLifecycle(Lifecycle.stable()).flatMap(map -> decode(ops, map)).map(r -> Pair.of(r, input));
+	}
+
+	@Override
+	public <T> DataResult<T> encode(final Map<K, V> input, final DynamicOps<T> ops, final T prefix) {
+		return encode(input, ops, ops.mapBuilder()).build(prefix);
+	}
+
+	/**
+	 * In {@link BaseMapCodec#decode(DynamicOps, MapLike)},
+	 * the whole deserialization will fail if one element fails.
+	 * `apply2stable` will return fail when any of the two elements is failed.
+	 * In this implementation, if one deserialization fails, it will log and ignore.
+	 */
+	@Override
+	public <T> DataResult<Map<K, V>> decode(final DynamicOps<T> ops, final MapLike<T> input) {
+		final ImmutableMap.Builder<K, V> builder = ImmutableMap.builder();
+
+		input.entries().forEach(pair -> {
+			try {
+				final DataResult<K> k = keyCodec().parse(ops, pair.getFirst());
+				final DataResult<V> v = elementCodec().parse(ops, pair.getSecond());
+
+				k.get().ifRight(kPartialResult -> {
+					LOGGER.error("Failed to decode key {} from {}  {}", k, pair, kPartialResult);
+				});
+				v.get().ifRight(vPartialResult -> {
+					LOGGER.error("Failed to decode value {} from {}  {}", v, pair, vPartialResult);
+				});
+
+				if (k.get().left().isPresent() && v.get().left().isPresent()) {
+					builder.put(k.get().left().get(), v.get().left().get());
+				} else {
+					// ignore failure
```

Comment:
Its a little unclear to me (with only basic knowlage of dfu) what failuire path is taken when a world is missing. May be a good idea to log out when this happens, or this is already handled by one of the other cases?

Reply:
The failure path is normally that, when it tries to parse a generator, it cannot recognize the generator type, then the failure propagates to the whole dimension information and the whole deserialization fails. In old versions it will still load the world but have the non-overworld dimensions removed. In newer versions it will show a screen that it fails to load and the user can click "safe mode", althought clicking "safe mode" doesn't do anything.
With this PR, when the generator fails to deserialize, it will be ignored.